client/src/main/scala/org/http4s/client/websocket/WSClient.scala

```diff
@@ -0,0 +1,179 @@
+  def sendPing(data: ByteVector = ByteVector.empty): F[Unit]
+
+  /** Send a Close frame. The sending side of this connection will be closed. */
+  def sendClose(reason: String = ""): F[Unit]
+
+  /** Wait for a websocket frame to be received. Returns `None` if the receiving side is closed.
+    * Fragmentation is handled automatically, the `last` attribute can be ignored.
+    */
+  def receive: F[Option[WSDataFrame]]
+
+  /** A stream of the incoming websocket frames. */
+  final def receiveStream: Stream[F, WSDataFrame] = Stream.repeatEval(receive).unNoneTerminate
+
+  /** The negotiated subprotocol, if any. */
+  def subprocotol: Option[String]
+
+  /** The close frame, if available. */
+  def closeFrame: TryableDeferred[F, WSFrame.Close]
+}
+
+trait WSClient[F[_]] {
+
+  /** Establish a websocket connection. It will be closed automatically if necessary. */
+  def connect(request: WSRequest): Resource[F, WSConnection[F]]
+
+  /** Establish a "high level" websocket connection. You only get to handle Text and Binary frames.
+    * Pongs will be replied automatically. Received frames are grouped by the `last` attribute. The
+    * connection will be closed automatically.
+    */
+  def connectHighLevel(request: WSRequest): Resource[F, WSConnectionHighLevel[F]]
+}
+
+object WSClient {
+  def defaultImpl[F[_]](
+      respondToPings: Boolean
+  )(f: WSRequest => Resource[F, WSConnection[F]])(implicit F: Concurrent[F]): WSClient[F] =
+    new WSClient[F] {
+      override def connect(request: WSRequest) = f(request)
+      override def connectHighLevel(request: WSRequest) =
+        for {
+          recvCloseFrame <- Resource.eval(Deferred.tryable[F, WSFrame.Close])
+          outputOpen <- Resource.eval(Ref[F].of(false))
+          conn <- f(request)
+        } yield new WSConnectionHighLevel[F] {
+          override def send(wsf: WSDataFrame) = conn.send(wsf)
+          override def sendMany[G[_]: Foldable, A <: WSDataFrame](wsfs: G[A]): F[Unit] =
+            conn.sendMany(wsfs)
+          override def sendPing(data: ByteVector) = conn.send(WSFrame.Ping(data))
+          override def sendClose(reason: String) =
+            conn.send(WSFrame.Close(1000, reason)) *> outputOpen.set(false)
+          override def receive: F[Option[WSDataFrame]] = {
+            def receiveDataFrame: OptionT[F, WSDataFrame] =
+              OptionT(conn.receive).flatMap { wsf =>
+                OptionT.liftF(wsf match {
+                  case WSFrame.Ping(data) if respondToPings => conn.send(WSFrame.Pong(data))
+                  case wsf: WSFrame.Close =>
+                    recvCloseFrame.complete(wsf) *> outputOpen.get.flatMap(conn.send(wsf).whenA(_))
+                  case _ => F.unit
+                }) >> (wsf match {
+                  case wsdf: WSDataFrame => OptionT.pure[F](wsdf)
+                  case _ => receiveDataFrame
+                })
+              }
+            def defrag(text: Chain[String], binary: ByteVector): OptionT[F, WSDataFrame] =
```

Comment:
IIRC the existing websocket parser in http4s already defragments messages for you (check me on this). should this API expect the underlying `WSConnection` to do that already or not? we can easily influence that with blaze/ember, but other backends may already perform it by the time we get the frames